refactor: feedback and improvements around the ShopifyModule

jmcdo29 opened the pull request:
I've addressed the majority of the feedback I provided to Andy a few days back, including creating a Shopify provider that can be injected, removing several unnecessary module imports, and removing all of the `forwardRef` uses which cleans up how the injections look significantly.

There's a few things I'll point out in comments that we'll probably want to address before merging this, but hopefully this can be improved on in other apps as well.

jmcdo29 commented on apps/shopify/src/config/shopify.config.joi.ts:
```diff
@@ -1,8 +1,8 @@
 import Joi from 'joi'
 
-export const shopifyConfigJoi = {
+export const shopifyConfigJoi = Joi.object({
```
I didn't remove the Joi config as mentioned in #86 yet due to wanting to work on the implementation for the `ConfigModule` extension we talked about over in #89. Once we have that created, and can validate the config on a per-module basis, I'll be happy to come back to this. But I did at least change it to a `Joi.object` so we can just use it directly instead of needing to `Joi.object(shopifyConfigJoi)`

andyslack replied:
Sounds good, lets circle back to this and replace as it makes sense to use Class Validator as this is what we use in the other parts of the system

jmcdo29 commented on apps/shopify/src/modules/orders/orders.controller.ts:
```diff
@@ -24,9 +24,8 @@ export class ShopifyOrdersController {
-		@Query('installed_app_id') installed_app_id: number,
 		@AccountId() account_id: number,
-		@Query() query: any,
+		@Query() { installed_app_id, ...query }: { installed_app_id: number; [key: string]: any },
```
I would like to create a specific type for this, but due to the `validationPipeOpitons` in `@juicyllama/core` using `whitelist: true` and `class-valdiator` not supporting the possibility of record properties (a class-valdiator and typescript language limitation) there's not much that can be done unless we uses `import type` and then a custom pipe (happy to implement it, but didn't want to make the decision without a discussion)

andyslack replied:
Hi @jmcdo29 sounds like a good conversation to have, to understand the pros/cons of different approaches, do you want to kickoff a discussion here: https://github.com/juicyllama/framework/discussions ?

jmcdo29 replied:
[Discussion created here](https://github.com/juicyllama/framework/discussions/94). I'm excited to see what the outcome of the conversation is

jmcdo29 reviewed (commented):
Most of the removals is getting rid of the `forwardRef` or organizing the imports alphabetically (aesthetic choice that I follow when possible). If there are any specific questions I'm more than happy to answer them.

andyslack commented:
Hi @jmcdo29 

This looks like great work and really nice to see a lot of code being removed and cleaned up. I am also a fan of alphabetically imports. 

A couple of observations:

**Testing**: I feel like we need some testing to make sure everything runs well before releasing, what do you think? Should we create some basic tests on the services just to ensure everything runs without error?

You can see how tests are working in the WordPress app for example: [WordPress Posts Service Test](https://github.com/juicyllama/framework/blob/main/apps/wordpress/src/modules/posts/wordpress.posts.service.test.spec.ts)

**Docs**: It maybe that the docs might need changing to reflect your changes? For example injecting the provider? Although I might be wrong? Can you take a look thought the [Shopify docs](https://docs.juicyllama.com/apps/shopify/getting-started) to see if anything needs to change?

jmcdo29 commented:
@andyslack I've seen this pattern before, and generally I like it as an approach. I've seen both a hybrid approach, where there's two or three definitive parameters and everything else is in the object that can have added properties, and I've seen just having an object by default. Both work well, but I think we'll want to make use of destructuring if we go with having the single object parameter, to make the usage of the properties more straightforward

jmcdo29 commented:
All right, I've added unit tests around most of the services in the shopify app, feel free to ask questions on them and how they work, and I've updated the documentation to show ow to use the `@InjectShopify()` decorator with the `ShopifyProviderModule`

jmcdo29 commented:
For the failing test, locally I added a `SHOPIFY_CLIENT_SECRET` and `SHOPIFY_CLIENT_ID` value to the `.env.development` so that it could run. I just added in junk values as the calls get mocked out anyways, but figured I shouldn't push up a change like that without bringing it up first

andyslack commented:
@jmcdo29 This all looks great to me.

@Yurko-Fedoriv Do you want to take a look to see if there is anything else you want to add before we publish this as our "template" for starting to build out more apps?

The only thing remaining is replacing Joi with Class Validator, which is blocked by #89 

We can circle back to this later. 

Once we are all happy, we can release this and @jmcdo29 if you want to pickup #89  and @Yurko-Fedoriv if you fancy picking up #73

andyslack commented:
I'm going to go ahead and merge this pull request so it prevents blocking from new apps being created with this template, we can continue to use this brand for further improvements/tweaks.

Yurko-Fedoriv commented:
Looks good.
The only note I have at this point, in tests I find it more useful to have
```ts
await expect(call()).resolves.toWhatever()
```
instead of
```ts
 expect(await call()).toWhatever()
```

The proposed variant will cause the failing test to show the failing test line instead of pointing out where the exception was thrown from.
